osu.Game.Rulesets.Mania/ManiaRuleset.cs

```diff
@@ -332,7 +332,15 @@ public override IEnumerable<KeyBinding> GetDefaultKeyBindings(int variant = 0)
             return Array.Empty<KeyBinding>();
         }
 
-        public override LocalisableString GetVariantName(int variant)
+        public override string GetVariantName(IBeatmap beatmap)
+        {
+            if (beatmap is ManiaBeatmap maniaBeatmap)
+                return GetVariantName(maniaBeatmap.OriginalTotalColumns);
+
+            return base.GetVariantName(beatmap);
```

Comment:
If stable is anything to go by, then this'll probably eventually need to somehow work for converts too, which is going to be a lot of fun to figure out for sure and probably shouldn't happen in this pull, but maybe it can be planned for early? Not sure the direction here is going to lend well to supporting this for converts as well...

Reply:
This is actually a very valid and blocking concern. I don't think the complexity in this PR (adding variant to the beatmap) is necessary at this point in time.

> An alternative solution to this is to process this at runtime, but I don't know how well that will scale.

Scale in terms of what? Rulesets can (and do) already have custom search strings for song select, so I wouldn't expect searching to also look at the variant name.
And I think if a ruleset comes up that needs significant amount more processing (i.e. can't be derived from `CircleSize` like mania) then it's probably better fit for beatmap statistics? Or at least if it must be a "variant" then that can probably be considered at that point in time.

Reply:
Hmm, so you're suggesting to shortcut it and just check `circle_size` in realm? The way I was doing it was to add to the real-time diffcalc pathway and use existing variant lookup method flow to avoid requiring any implementation overhead.

Reply:
I don't know why it has to be checked in realm in the first place. It's a purely display component.

https://osu.ppy.sh/beatmapsets/18260#taiko/65313
```
database: 6b4551ebd46058dd31ff42565cf7d18c
osu: 7ce1ab636a0b118ee0c76cda4b2e1818
archive: 6b4551ebd46058dd31ff42565cf7d18c
```

https://osu.ppy.sh/beatmapsets/758986#fruits/1596618
```
database: 2c46f40a654e9741078219c0762d1cc7
osu: f101011c6e3cd26a987eb8309c992b09
archive: 2c46f40a654e9741078219c0762d1cc7
```

But I guess it follows the same pattern - so both the archive and database checksums are probably incorrect?

Reply:
Hmm, I thought we already had code for mania which extracted the key count statistics from beatmaps, but it turns out that's done on a case-by-case basis....
https://github.com/ppy/osu/blob/ccbac08985df5819330ab2b3537f5c9d08dc5345/osu.Game/Screens/Select/Details/AdvancedStats.cs#L121-L134

Dunno. I suppose it could be fine to chuck it in diffcalc.